src/cjss.js

```diff
@@ -1,93 +1,73 @@
-import getPureProperty from './getPureProperty';
-import functionFromString from './functionFromString';
+import getProperty from './getProperty';
 import ruleList from './ruleList';
+import { getPlugin } from './registerPlugin';
+import Stage from './Stage';
+import CJSSError from './CJSSError';
 
 /**
  * Run one CJSS rule, handling the properties `--html`, `--js` and `--data`.
  *
  * @param {CSSRule} rule The rule to parse.
- * @returns {Boolean} Whether the operation was successful.
  */
 function processRule(rule) {
   const selector = rule.style.parentRule.selectorText;
-  const elements = document.querySelectorAll(selector);
+  const elements = selector === 'script' ? [null] : document.querySelectorAll(selector);
 
-  const js = getPureProperty(rule, '--js');
-  const html = getPureProperty(rule, '--html');
-  let data = getPureProperty(rule, '--data');
-
-  try {
-    data = JSON.parse(`{${ data }}`);
-  } catch (e) {
-    if (e instanceof SyntaxError) {
-      console.error(`CJSS: Invalid JSON found at ${ selector }: {${ data }}`);
-      console.error(e.message);
-      return false;
-    } else throw e;
-  }
-
-  if (html) {
-    for (const element of elements) {
-      const code = `return (\`${ html }\`)`;
-      const render = functionFromString(code, ['data', 'yield']);
+  const compiledStages = [];
 
+  for (const stage of Stage.ordered) {
+    const property = getProperty(rule, stage)
+      || (stage.fallback && getProperty(rule, stage.fallback));
+    if (property) {
+      const { mode, body } = property;
       try {
-        element.innerHTML = render.run([data, element.innerHTML]);
+        compiledStages.push({
+          body, mode, stage, run: getPlugin(stage, mode)(body),
+        });
       } catch (e) {
-        console.error('CJSS: Error in HTML:', e);
-        console.error(`at selector '${ selector }' and element`, element);
-        console.error(`of script:\n${ js }`);
-        return;
+        if (e instanceof CJSSError) {
+          console.error(e.message, e.original);
+          console.error(`in script ${stage}: ${mode}(${body})`);
+          // TODO: in the case of an error, should we skip this element or every element?
```

Comment:
After hitting an error when compiling a rule, is logging the error and moving on to the next rule good enough?